presto-main/src/main/java/com/facebook/presto/execution/buffer/SpoolingOutputBuffer.java

```diff
@@ -0,0 +1,761 @@
+    private final AtomicLong totalBufferedPages = new AtomicLong();
+    private final AtomicLong totalPagesAdded = new AtomicLong();
+    private final AtomicLong totalRowsAdded = new AtomicLong();
+
+    private final OutputBufferId outputBufferId = new OutputBufferId(0);
+    private final Closer closer = Closer.create();
+
+    private static ExecutorService coreExecutor = newCachedThreadPool(daemonThreadsNamed("spooling-outputbuffer-%s"));
+    private static final Logger log = Logger.get(SpoolingOutputBuffer.class);
+
+    private final AtomicBoolean noMorePages = new AtomicBoolean();
+    private final AtomicLong currentMemorySequenceId = new AtomicLong();
+    private final AtomicLong currentSequenceId = new AtomicLong();
+    private final AtomicLong startPage = new AtomicLong();
+    private final AtomicLong totalPagesRemaining = new AtomicLong();
+
+    @GuardedBy("this")
+    private final LinkedList<HandleInfo> handleInfoQueue = new LinkedList<>();
+
+    @GuardedBy("this")
+    private final Queue<SerializedPage> pages = new ArrayDeque<>();
+
+    @GuardedBy("this")
+    private PendingRead pendingRead;
+
+    public SpoolingOutputBuffer(
+            TaskId taskId,
+            String taskInstanceId,
+            OutputBuffers outputBuffers,
+            StateMachine<BufferState> state,
+            TempStorage tempStorage,
+            DataSize threshold)
+    {
+        requireNonNull(taskId, "taskId is null");
+        this.taskInstanceId = requireNonNull(taskInstanceId, "taskInstanceIs is null");
+        this.outputBuffers = requireNonNull(outputBuffers, "outputBuffers is null");
+        this.state = requireNonNull(state, "state is null");
+        this.tempStorage = requireNonNull(tempStorage, "tempStorage is null");
+        this.threshold = requireNonNull(threshold, "threshold is null");
+
+        context = new TempDataOperationContext(Optional.empty(), taskId.getQueryId().toString(), Optional.empty(), new Identity("spooling-buffer", Optional.empty()));
+
+        closer.register(coreExecutor::shutdownNow);
+        executor = listeningDecorator(coreExecutor);
+
+        state.compareAndSet(OPEN, NO_MORE_BUFFERS);
+        state.compareAndSet(NO_MORE_PAGES, FLUSHING);
+        checkFlushComplete();
+    }
+
+    @Override
+    public OutputBufferInfo getInfo()
+    {
+        return new OutputBufferInfo(
+                "SPOOLING",
+                state.get(),
+                state.get().canAddBuffers(),
+                state.get().canAddPages(),
+                totalBufferedBytes.get(),
+                totalBufferedPages.get(),
+                totalRowsAdded.get(),
+                totalPagesAdded.get(),
+                ImmutableList.of()
+        );
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return state.get() == BufferState.FINISHED;
+    }
+
+    @Override
+    public double getUtilization()
+    {
+        return 0;
```

Comment:
Good point. Originally, I planned to have everything written to TempStorage, so it would be always 0. Now that we're keeping pages in memory, it'll be non zero.